### PR TITLE
Include generic type parameters in .d.ts clutz aliases.

### PR DESCRIPTION
**Pull request by LucasSloan**
Adds support and a test of generic type parameters in .d.ts clutz aliases.

### CODE REVIEW

File: src/tsickle.ts [hunk 1]
```diff
@@ -2079,6 +2084,56 @@ function addClutzAliases(
...
+function getGenericTypeParameters(symbol: ts.Symbol): {params: string, constrainedParams: string} {
+  if (!symbol.declarations) {
+    return {params: '', constrainedParams: ''};
+  }
+
+  if (!symbol.declarations[0]) {
+    return {params: '', constrainedParams: ''};
+  }
+
+  const declaration = symbol.declarations[0];
```
Comment (rkirov) [inline src/tsickle.ts]: add a comment about the assumption we talked about here 'all declarations have to have matching generic types'

File: src/tsickle.ts [hunk 2]
```diff
@@ -2079,6 +2084,56 @@ function addClutzAliases(
...
+  if (!declarationWithTypeParameters.typeParameters) {
+    return {params: '', constrainedParams: ''};
+  }
+
+  const paramList: string[] = [];
+  const constrainedParamList: string[] = [];
+  for (const param of declarationWithTypeParameters.typeParameters) {
+    if (param.constraint) {
+      constrainedParamList.push(`${param.name.text} extends ${
+          (param.constraint as ts.TypeReferenceNode).typeName.getText()}`);
```
Comment (rkirov) [inline src/tsickle.ts]: This cast is not safe - consider `class C<T extends {a: string}>`. Technically the bound could be any type expression (not just reference). Why can't we just do ` param.constraint.getText()`.

File: src/tsickle.ts [hunk 3]
```diff
@@ -2079,6 +2084,56 @@ function addClutzAliases(
...
+    } else {
+      constrainedParamList.push(param.name.text);
```
Comment (rkirov) [inline src/tsickle.ts]: .text vs .getText() is there a difference?

File: src/tsickle.ts [hunk 4]
```diff
@@ -2050,18 +2050,23 @@ function addClutzAliases(
       // https://github.com/angular/clutz/tree/master/src/test/java/com/google/javascript/clutz
       const clutzSymbolName = 'module$contents$' + clutzModuleName + '_' + symbol.name;
 
+      const genericParams = getGenericTypeParameters(symbol);
+
       if (symbol.flags & ts.SymbolFlags.Class) {
         // classes need special care to match clutz, which seperates class types into a
         // type for the static properties and a type for the instance properties
-        reexports += `\t\ttype ${clutzSymbolName} = ${symbol.name};\n`;
+        reexports += `\t\ttype ${clutzSymbolName}${genericParams.constrainedParams} = ${
```
Comment (rkirov) [inline src/tsickle.ts]: This is getting gnarly, consider helper methods like emit, indent,unindent. `https://github.com/angular/clutz/blob/master/src/main/java/com/google/javascript/clutz/DeclarationGenerator.java#L1129`.

Could be done in the next PR.

File: src/tsickle.ts [hunk 5]
```diff
@@ -2079,6 +2084,56 @@ function addClutzAliases(
   return dtsFileContent;
 }
 
+/**
+ * Returns 2 strings specifying the generic type arguments for the symbol.  The constrained params
+ * include any `T extends foo` arguments, the regular params are just a list of the type symbols,
+ * since we need the constraints on the LHS of the alias declaration, but can't have them on the
+ * RHS.
+ */
+function getGenericTypeParameters(symbol: ts.Symbol): {params: string, constrainedParams: string} {
```
Comment (rkirov) [inline src/tsickle.ts]: I think 'paramsWithContraint' is a better name. For awhile, I thought you split the params into two buckets, ones with and without contraints.

File: src/tsickle.ts [hunk 6]
```diff
@@ -2079,6 +2084,56 @@ function addClutzAliases(
...
+    if (param.constraint) {
+      constrainedParamList.push(`${param.name.text} extends ${
+          (param.constraint as ts.TypeReferenceNode).typeName.getText()}`);
+    } else {
+      constrainedParamList.push(param.name.text);
+    }
+    paramList.push(param.name.getText());
+  }
+
+  const params = `<${paramList.join(',')}>`;
```
Comment (rkirov) [inline src/tsickle.ts]: nit, I think the < > and ',' belong in the emit code, and this helper should emit {param: string[], paramWithConstrain: string[]}

File: src/tsickle.ts [hunk 7]
```diff
@@ -2050,18 +2050,23 @@ function addClutzAliases(
       // https://github.com/angular/clutz/tree/master/src/test/java/com/google/javascript/clutz
       const clutzSymbolName = 'module$contents$' + clutzModuleName + '_' + symbol.name;
 
+      const genericParams = getGenericTypeParameters(symbol);
```
Comment (rkirov) [inline src/tsickle.ts]: `const {params, constrainedParams}`

File: src/tsickle.ts (hunk 3 above)
Comment (LucasSloan) [reply src/tsickle.ts]: Some things have both and some just have the `.getText()`.  I've standardized on `.getText()`.

File: src/tsickle.ts (hunk 6 above)
Comment (LucasSloan) [reply src/tsickle.ts]: I had it that way originally - I don't like how that requires additional variables, or a ternary expression at every usage.

File: src/tsickle.ts (hunk 4 above)
Comment (LucasSloan) [reply src/tsickle.ts]: Made a partial step in that direction.

File: src/tsickle.ts (hunk 2 above)
Comment (LucasSloan) [reply src/tsickle.ts]: Done.

File: src/tsickle.ts [hunk 8]
```diff
@@ -2050,35 +2050,95 @@ function addClutzAliases(
...
   }
 
   return dtsFileContent;
 }
 
+/**
+ * Returns 2 strings specifying the generic type arguments for the symbol.  The constrained params
+ * include any `T extends foo` arguments, the regular params are just a list of the type symbols,
+ * since we need the constraints on the LHS of the alias declaration, but can't have them on the
+ * RHS.
```
Comment (rkirov) [inline src/tsickle.ts]: add that we include default here too.

**Pull request merged**